enclave/core/sgx/once.c

```diff
@@ -5,34 +5,78 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define FUNC_NOT_INVOKED 0
+#define FUNC_BEING_INVOKED 1
+#define FUNC_INVOKED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
-    OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
+
+    oe_once_t status = __atomic_load_n(once, __ATOMIC_ACQUIRE);
+
+    /*
+      Use an atomic-acquire load operation to check if the function has not been
+      invoked. If the function has already been invoked, there is nothing to do.
+      If the function is being invoked, then this thread must wait for the
+      function invocation to complete. Otherwise, this thread can try to take
+      ownership of invoking the function
+    */
+    if (status != FUNC_INVOKED)
     {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+        /*
+          Multiple threads could reach here simultaneously after checking
+          whether the function has been invoked or not. Only one of them must
+          now invoke the function and others must wait for the function
+          invocation to complete.
+          To determine who gets to invoke the function, each thread atomically
+          compares the value of once to FUNC_NOT_INVOKED and if equal, sets the
+          value to FUNC_BEING_INVOKED to signal to other threads that the
+          function is being invoked.
+
+          If the compare and exchange succeeds, then this thread owns the
+          responsibility of invoking the function.
 
-        oe_spin_lock(&_lock);
+          If the compar and exchange fails, then another thread has taken
+          ownership of calling the function and therefore this thread must
+          wait for the other thread to complete the invocation.
 
-        if (*once == 0)
+        */
+        oe_once_t expected = FUNC_NOT_INVOKED;
+        bool retval = __atomic_compare_exchange_n(
+            once,
+            &expected, // If the current value of once if FUNC_NOT_INVOKED
+            FUNC_BEING_INVOKED, // take ownership of calling the function.
+            false,              // Use a strong compare exchange.
+            __ATOMIC_RELEASE,   // If exchange was successful, use release
```

Comment:
I would use __ATOMIC_ACQ_REL for the success case and __ATOMIC_ACQUIRE for the fail case.  That's my only nit.  The rest of this code LGTM.